pkg/api/entries.go

```diff
@@ -272,54 +272,37 @@ func getEntryURL(locationURL url.URL, uuid string) strfmt.URI {
 
 // GetLogEntryByUUIDHandler gets log entry and inclusion proof for specified UUID aka merkle leaf hash
 func GetLogEntryByUUIDHandler(params entries.GetLogEntryByUUIDParams) middleware.Responder {
-	ctx := params.HTTPRequest.Context()
-
 	uuid, err := sharding.GetUUIDFromIDString(params.EntryUUID)
 	if err != nil {
 		return handleRekorAPIError(params, http.StatusBadRequest, err, "")
 	}
-	var tid int64
 	tidString, err := sharding.GetTreeIDFromIDString(params.EntryUUID)
-	if err != nil {
-		// If EntryID is plain UUID, assume no sharding and use ActiveTreeID. The ActiveTreeID
-		// will == the tlog_id if a tlog_id is passed in at server startup.
-		if err.Error() == "cannot get treeID from plain UUID" {
-			tid = api.logRanges.ActiveTreeID()
-		} else {
-			return handleRekorAPIError(params, http.StatusBadRequest, err, "")
-		}
-	} else {
-		tid, err = strconv.ParseInt(tidString, 16, 64)
+
+	// If treeID is found in EntryID, route to correct tree
+	if err == nil {
+		tid, err := strconv.ParseInt(tidString, 16, 64)
 		if err != nil {
 			return handleRekorAPIError(params, http.StatusBadRequest, err, "")
 		}
-	}
-	hashValue, _ := hex.DecodeString(uuid)
-
-	tc := NewTrillianClientFromTreeID(params.HTTPRequest.Context(), tid)
-	log.RequestIDLogger(params.HTTPRequest).Debugf("Retrieving UUID %v from TreeID %v", uuid, tid)
-
-	resp := tc.getLeafAndProofByHash(hashValue)
-	switch resp.status {
-	case codes.OK:
-	case codes.NotFound:
-		return handleRekorAPIError(params, http.StatusNotFound, fmt.Errorf("grpc error: %w", resp.err), "")
-	default:
-		return handleRekorAPIError(params, http.StatusInternalServerError, fmt.Errorf("grpc error: %w", resp.err), trillianUnexpectedResult)
+		resp, _ := RetrieveUUID(params, uuid, tid)
+		return resp
 	}
 
-	result := resp.getLeafAndProofResult
-	leaf := result.Leaf
-	if leaf == nil {
-		return handleRekorAPIError(params, http.StatusNotFound, errors.New("grpc returned 0 leaves with success code"), "")
-	}
+	// If EntryID is plain UUID (ex. from client v0.5), check all trees
+	if err.Error() == "cannot get treeID from plain UUID" {
```

Comment:
Best practice would be not to hardcode this string in case someone changes the wording on the error later on. You could either:

1. Create an error type to check for ([example here](https://earthly.dev/blog/golang-errors/#defining-expected-errors) which uses `errors.Is`)
2. Change the `GetUUIDFromIDString` function to return `"", nil` in the case that there is no treeID. That way the error is nil, but we know if the treeID is empty that we have to go through all the shards

Reply:
Yes, these options make more sense.